src/modules/BootStorage.rb

```diff
@@ -142,31 +142,48 @@ def stage1_device_for_name(dev_name)
     def stage1_partitions_for(device)
       # so how to do search? at first find first partition with parents
       # that is on disk or multipath (as ancestors method is not sorted)
-      to_process = [device]
-      partitions = []
-      loop do
-        break if to_process.empty?
-        partition = to_process.pop
-        if partition.is?(:partition)
-          partitionable = partition.partitionable
-          if partitionable.is?(:disk) || partitionable.is?(:multipath)
-            partitions << partition
-            next # we are done here, we found partition for this part
-          end
+      partitions = find_ancestors(device) do |ancestor|
+        if ancestor.is?(:partition)
+          partitionable = ancestor.partitionable
+          partitionable.is?(:disk) || partitionable.is?(:multipath)
+        else
+          false
         end
-        to_process.concat(partition.parents)
       end
 
       # now replace all logical partitions for extended
       partitions.map! { |p| extended_for_logical(p) }
       partitions.uniq!
 
-      log.info "stage1 partitions for #{device.inspect} is #{partitions.inspect}"
+      log.info "stage1 partitions for #{device.inspect} are #{partitions.inspect}"
 
       partitions
     end
 
-    # If passed partition is logical, return extended, otherwise return argument
+    # In a device graph, starting at *device* (inclusive), find the parents
+    # that match *predicate*.
+    # NOTE that once the predicate matches, the search stops **for that node**
+    # but continues for other nodes.
+    # @param device [Y2Storage::Device] starting point
+    # @yieldparam [Y2Storage::Device]
+    # @yieldreturn [Boolean]
+    # @return [Array<Y2Storage::Device]
+    def find_ancestors(device, &predicate)
```

Comment:
better name will be `select_ancestors`

Reply:
and second note, as it is internal helper only, please move it to internal private section of class.